### ISSUE TITLE
Datum-weighted loss functions

### ISSUE THREAD

**Issue by jj-globant** [OpenNMT/OpenNMT-py#271]
My use case is as follows:
We are using a _translation_ of sorts to build a chatbot, and we have different sources for the question-answer pairs which have a different degree of _trustworthiness_.

I'm looking into being able to use a value that represents how authoritative a single datum is, to weight the calculation of the loss and affect less the ones which i want to use for the sake of completeness but that shouldn't be as important as the most authoritative ones.

In order to do so, i would need to extend / modify de `DatumWeightedNMTCriterion` class so that instead of using `nn.NLLLoss` as `crit`, it would use something similar but _datum-weighted_.

- Is there anything in the roadmap that looks like this?
- I'll be working on it anyway on our fork for our use case, but the ideal would be to contribute to the project. In that sense:
  - How should i structure it to be more in line with styles and conventions of the proyect?
  - This is what i'm thinking of:
    - extend `onmt/IO.py` to be able to add a property to the targets (the weight), since it's only usefull in training, and it's closely paired with it's value
    - extend `onmt/Loss.py` so that i would return a different `crit` form the actual, `crit = nn.NLLLoss(weight, size_average=False)`
    - if there are no _datum-weights_, it would return `crit = nn.NLLLoss(weight, size_average=False, datum_weights=False)`
    - adapt `Statistics` and `LossCompute` to handle the extra argument when it exists, but wihtout modifying current behavior
     - new `crit` would very closely follow `torch/nn/_functions/thnn/auto_double_backwards.py`, which is the computation that gets used on `NMTCriterion` in the end.
  - Another option is to create a new module `datum_weights` where the modified variants could be found, so it could be tested as experimental before thinking about adding it to the project.

Thoughts and ideas are very welcome!

**Comment by srush** [comment]
This sounds great! We would love a PR. Your design seems good, weights become a field on batch and get passed to the criterion.

I agree the structure of loss compute is somewhat brittle. You shouldn't have to add it manually to loss compute and statistics. As you build keep in mind how you might generalize and simplify these.

Also remember that in pytorch criterion are not so special. If you can write your objective with auto grad then you don't need an explicit criterion at all.
